src/rust/common/src/expiry.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use config::TimeType;
+use rustcommon_time::{CoarseDuration, Duration};
+
+use std::time::SystemTime;
+
+pub struct Expiry {
+    time_type: TimeType,
+    expiry: u32,
+}
+
+impl Expiry {
+    pub fn new(expiry: u32, time_type: TimeType) -> Self {
+        Self { expiry, time_type }
+    }
+
+    // TODO(bmartin): this conversion can be made more efficient
+    pub fn as_secs(&self) -> u32 {
+        match self.time_type {
+            TimeType::Unix => {
+                let epoch = SystemTime::now()
+                    .duration_since(SystemTime::UNIX_EPOCH)
+                    .unwrap()
+                    .as_secs() as u32;
+                self.expiry.wrapping_sub(epoch)
+            }
+            TimeType::Delta => self.expiry,
+            TimeType::Memcache => {
+                if self.expiry < 60 * 60 * 24 * 30 {
+                    self.expiry
+                } else {
+                    let epoch = SystemTime::now()
+                        .duration_since(SystemTime::UNIX_EPOCH)
+                        .unwrap()
+                        .as_secs() as u32;
+                    self.expiry.wrapping_sub(epoch)
```

Comment:
I don't think `wrapping_sub` is what we need? If someone gives a timestamp that's older than the now epoch, we want it to immediately expire, not give it a large value. So essentially we want to collapse all negative values to 0.

The C implementation didn't have to deal with this explicitly because everything was converted to process time, so the comparison with `now` was done when the keys are later accessed.